balancer/rls/internal/keys/builder_test.go

```diff
@@ -0,0 +1,321 @@
+		wantErr        bool
+		wantErrPrefix  string
+	}{
+		{
+			desc:           "No GrpcKeyBuilder",
+			cfg:            &rlspb.RouteLookupConfig{},
+			wantBuilderMap: nil,
+			wantErr:        true,
+			wantErrPrefix:  "rls: RouteLookupConfig does not contain any GrpcKeyBuilder",
+		},
+		{
+			desc: "Two GrpcKeyBuilders with same Name",
+			cfg: &rlspb.RouteLookupConfig{
+				GrpcKeybuilders: []*rlspb.GrpcKeyBuilder{goodKeyBuilder1, goodKeyBuilder1},
+			},
+			wantBuilderMap: nil,
+			wantErr:        true,
+			wantErrPrefix:  "rls: GrpcKeyBuilder in RouteLookupConfig contains repeated Name field",
+		},
+		{
+			desc: "GrpcKeyBuilder with empty Service field",
+			cfg: &rlspb.RouteLookupConfig{
+				GrpcKeybuilders: []*rlspb.GrpcKeyBuilder{emptyServiceKeyBuilder, goodKeyBuilder1},
+			},
+			wantBuilderMap: nil,
+			wantErr:        true,
+			wantErrPrefix:  "rls: GrpcKeyBuilder in RouteLookupConfig contains a Name field with no Service",
+		},
+		{
+			desc: "GrpcKeyBuilder with no Name",
+			cfg: &rlspb.RouteLookupConfig{
+				GrpcKeybuilders: []*rlspb.GrpcKeyBuilder{{}, goodKeyBuilder1},
+			},
+			wantBuilderMap: nil,
+			wantErr:        true,
+			wantErrPrefix:  "rls: GrpcKeyBuilder in RouteLookupConfig does not contain any Name",
+		},
+		{
+			desc: "GrpcKeyBuilder with requiredMatch field set",
+			cfg: &rlspb.RouteLookupConfig{
+				GrpcKeybuilders: []*rlspb.GrpcKeyBuilder{requiredMatchKeyBuilder, goodKeyBuilder1},
+			},
+			wantBuilderMap: nil,
+			wantErr:        true,
+			wantErrPrefix:  "rls: GrpcKeyBuilder in RouteLookupConfig has required_match field set",
+		},
+		{
+			desc: "GrpcKeyBuilder two headers with same key",
+			cfg: &rlspb.RouteLookupConfig{
+				GrpcKeybuilders: []*rlspb.GrpcKeyBuilder{repeatedHeadersKeyBuilder, goodKeyBuilder1},
+			},
+			wantBuilderMap: nil,
+			wantErr:        true,
+			wantErrPrefix:  "rls: GrpcKeyBuilder in RouteLookupConfig contains repeated Key field in headers",
+		},
+		{
+			desc: "One good GrpcKeyBuilder",
+			cfg: &rlspb.RouteLookupConfig{
+				GrpcKeybuilders: []*rlspb.GrpcKeyBuilder{goodKeyBuilder1},
+			},
+			wantBuilderMap: wantBuilderMap1,
+		},
+		{
+			desc: "Two good GrpcKeyBuilders",
+			cfg: &rlspb.RouteLookupConfig{
+				GrpcKeybuilders: []*rlspb.GrpcKeyBuilder{goodKeyBuilder1, goodKeyBuilder2},
+			},
+			wantBuilderMap: wantBuilderMap2,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.desc, func(t *testing.T) {
+			builderMap, err := MakeBuilderMap(test.cfg)
+			if (err != nil) != test.wantErr {
+				t.Errorf("MakeBuilderMap(%+v) returned err: %v, want: %v", test.cfg, err, test.wantErr)
+			}
+			if err != nil && !strings.HasPrefix(err.Error(), test.wantErrPrefix) {
+				t.Errorf("MakeBuilderMap(%+v) returned err: %v, want: %v", test.cfg, err, test.wantErr)
+			}
+			if !cmp.Equal(builderMap, test.wantBuilderMap, cmp.AllowUnexported(builder{}, matcher{})) {
+				t.Errorf("MakeBuilderMap(%+v) returned %v, want: %v", test.cfg, builderMap, test.wantBuilderMap)
```

Comment:
Where possible, try to combine your checks so you only need one t.Errorf line:

```
if !strings.HasPrefix(<err>) || !cmp.Equal(<builderMap>) {
	t.Errorf("MakeBuilderMap(%+v) = %v, %v; want %v, HasPrefix(%v)", ...)  // e.g.
}

Reply:
Done.